File: src/OrderManager.Api/Controllers/InventoryController.cs
using Microsoft.AspNetCore.Mvc;
using OrderManager.Api.Services;

namespace OrderManager.Api.Controllers;

/// <summary>
/// API controller for querying and managing product inventory levels.
/// </summary>
[ApiController]
[Route("api/[controller]")]
public class InventoryController : ControllerBase
{
    private readonly InventoryService _inventoryService;

    /// <summary>
    /// Initializes a new instance of <see cref="InventoryController"/>.
    /// </summary>
    /// <param name="inventoryService">The service handling inventory business logic.</param>
    public InventoryController(InventoryService inventoryService)
    {
        _inventoryService = inventoryService;
    }

    /// <summary>
    /// Returns all inventory items with their associated product details.
    /// </summary>
    /// <returns>A 200 OK response containing a list of all inventory records.</returns>
    [HttpGet]
    public async Task<IActionResult> GetAll() => Ok(await _inventoryService.GetAllInventoryAsync());

    /// <summary>
    /// Returns the inventory record for a specific product.
    /// </summary>
    /// <param name="productId">The identifier of the product to look up.</param>
    /// <returns>A 200 OK response with the inventory item, or 404 Not Found if no record exists for the product.</returns>
    [HttpGet("product/{productId}")]
    public async Task<IActionResult> GetByProduct(int productId)
    {
        var item = await _inventoryService.GetInventoryByProductIdAsync(productId);
        return item is null ? NotFound() : Ok(item);
    }

    /// <summary>
    /// Adds stock to a product's inventory.
    /// </summary>
    /// <param name="productId">The identifier of the product to restock.</param>
    /// <param name="request">The payload containing the quantity to add.</param>
    /// <returns>A 200 OK response with the updated inventory item.</returns>
    [HttpPost("product/{productId}/restock")]
    public async Task<IActionResult> Restock(int productId, [FromBody] RestockRequest request)
    {
        var item = await _inventoryService.RestockAsync(productId, request.Quantity);
        return Ok(item);
    }

    /// <summary>
    /// Returns all inventory items whose on-hand quantity is at or below their reorder level.
    /// </summary>
    /// <returns>A 200 OK response containing a list of low-stock inventory records.</returns>
    [HttpGet("low-stock")]
    public async Task<IActionResult> GetLowStock() => Ok(await _inventoryService.GetLowStockItemsAsync());
}

/// <summary>Request payload for restocking a product's inventory.</summary>
/// <param name="Quantity">The number of units to add to the current stock.</param>
public record RestockRequest(int Quantity);

File: server/src/main/java/com/ordermanager/api/controller/InventoryController.java
package com.ordermanager.api.controller;

import com.ordermanager.api.dto.RestockRequest;
import com.ordermanager.api.model.InventoryItem;
import com.ordermanager.api.service.InventoryService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/api/inventory")
public class InventoryController {

    private final InventoryService inventoryService;

    public InventoryController(InventoryService inventoryService) {
        this.inventoryService = inventoryService;
    }

    @GetMapping
    public ResponseEntity<List<InventoryItem>> getAll() {
        return ResponseEntity.ok(inventoryService.getAllInventory());
    }

    @GetMapping("/product/{productId}")
    public ResponseEntity<InventoryItem> getByProduct(
            @PathVariable Integer productId) {
        return inventoryService.getInventoryByProductId(productId)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping("/product/{productId}/restock")
    public ResponseEntity<InventoryItem> restock(
            @PathVariable Integer productId,
            @RequestBody RestockRequest request) {
        InventoryItem item = inventoryService.restock(
                productId, request.getQuantity());
        return ResponseEntity.ok(item);
    }

    @GetMapping("/low-stock")
    public ResponseEntity<List<InventoryItem>> getLowStock() {
        return ResponseEntity.ok(inventoryService.getLowStockItems());
    }
}
